src/go/rpk/pkg/config/config.go

```diff
@@ -355,6 +355,26 @@ func checkRpkConfig(v *viper.Viper) []error {
 	return errs
 }
 
+func decoderConfig() mapstructure.DecoderConfig {
+	return mapstructure.DecoderConfig{
+		// Sometimes viper will save int values as strings (i.e.
+		// through BindPFlag) so we have to allow mapstructure
+		// to cast them.
+		WeaklyTypedInput: true,
+		DecodeHook: mapstructure.ComposeDecodeHookFunc(
+			v21_1_4MapToNamedSocketAddressSlice,
+		),
+	}
+}
+
+func decoderConfigOptions() viper.DecoderConfigOption {
+	return func(c *mapstructure.DecoderConfig) {
+		cfg := decoderConfig()
+		c.DecodeHook = cfg.DecodeHook
```

Comment:
This appears to reset any existing `DecodeHook`s, `Viper` sets a couple. I think it's worth composing if the original is not empty, or, probably better, we set out default above, to include them:
```go
mapstructure.ComposeDecodeHookFunc(
    mapstructure.StringToTimeDurationHookFunc(),
    mapstructure.StringToSliceHookFunc(","),
    v21_1_4MapToNamedSocketAddressSlice,
)
```

Reply:
Done!